src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+            return false;
+    } else
+        device->trigger_name = name;
+
+    return check_manual_triggering(device);
+}
+
+static void
+set_buffer_size(struct sol_iio_device *device, int buffer_size)
+{
+    char path[PATH_MAX + 1];
+
+    if (craft_filename_path(path, sizeof(path), BUFFER_LENGHT_DEVICE_PATH, device->device_id)) {
+        SOL_WRN("Could not set IIO device buffer size");
+        return;
+    }
+
+    if (!write_file(path, "%d", buffer_size)) {
+        SOL_WRN("Could not set IIO device buffer size");
+    }
+
+    return;
+}
+
+static bool
+set_buffer_enabled(struct sol_iio_device *device, bool enabled)
+{
+    char path[PATH_MAX + 1];
+    int len;
+
+    len = snprintf(path, sizeof(path), BUFFER_ENABLE_DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX)
+        return false;
+
+    if (!write_file(path, "%d", enabled))
+        return false;
+
+    return true;
+}
+
+static size_t
+calc_buffer_size(struct sol_iio_device *device)
+{
+    struct sol_iio_channel *channel;
+    int i;
+    size_t size = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&device->channels, channel, i) {
+        size += channel->storagebits;
+    }
+
+    return size / 8;
+}
+
+static bool
+device_reader_cb(void *data, int fd, unsigned int active_flags)
+{
+    struct sol_iio_device *device = data;
+    bool result = true;
+    size_t len = calc_buffer_size(device);
+    uint8_t *buffer = NULL;
+
+    buffer = calloc(1, len);
+    if (!buffer) {
+        SOL_WRN("Could not allocate memory for read buffer");
+        result = false;
+    }
+
+    if (result && active_flags & (SOL_FD_FLAGS_ERR | SOL_FD_FLAGS_HUP | SOL_FD_FLAGS_NVAL)) {
+        SOL_WRN("Unexpected reading");
+        result = false;
+    }
+
+    while (result) {
+        ssize_t ret;
+
+        ret = read(fd, buffer, len);
+        if (ret < 0) {
+            if (errno == EINTR)
+                continue;
+            else if (errno == EAGAIN || errno == EWOULDBLOCK)
+                break;
+            else {
+                result = false;
+                break;
+            }
+        }
+
+        if ((size_t)ret == len) {
+            if (device->reader_cb) {
+                struct sol_iio_buffer iio_buffer;
+
+                iio_buffer.size = len;
+                iio_buffer.data = buffer;
+                device->reader_cb(device, &iio_buffer, device->reader_cb_data);
+            }
+        }
+    }
+
+    free(buffer);
+
+    return result;
+}
+
+static bool
+setup_device_reader(struct sol_iio_device *device)
+{
+    int len, fd;
+    char path[PATH_MAX + 1];
+
+    len = snprintf(path, sizeof(path), DEVICE_PATH, device->device_id);
+    if (len < 0 || len > PATH_MAX) {
+        SOL_WRN("Could not open IIO device%d: Device does not exist", device->device_id);
+        return false;
+    }
+
+    fd = open(path, O_RDONLY | O_CLOEXEC | O_NONBLOCK);
+    if (fd < 0) {
+        SOL_WRN("Could not open IIO device%d: Device does not exist", device->device_id);
```

Comment:
it's better to %s and use `path` in this error message, followed by sol_util_strerrnoa(). It may be the case that the device does exist, but you don't have permission and the error message will be misleading.